### PR TITLE
IC-1063 to 1065, IC-1033: A probation practitioner can filter the list of intervention search results

### PR DESCRIPTION
**Pull request by lawrence-forooghian**
## What does this pull request do?

Adds filtering controls to the list of intervention search results, and a summary of the filters chosen.

## What is the intent behind these changes?

To allow a probation practitioner to narrow down the search results.

## Screenshot

image

### CODE REVIEW

File: server/routes/findInterventions/interventionsFilter.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
+import { Request } from 'express'
+import { InterventionsFilterParams } from '../../services/interventionsService'
+
+export default class InterventionsFilter {
+  pccRegionIds: string[] | undefined
+
+  gender: ('male' | 'female')[] | undefined
```
Comment (Gweaton) [inline server/routes/findInterventions/interventionsFilter.ts:7]: No biggie (and something we can do after this is merged if you think it's a good idea), but could be worth storing these in an exported enum, so we could use them all over the codebase?

File: server/routes/findInterventions/searchResultsView.ts [hunk 2]
```diff
@@ -5,14 +5,62 @@ import { SummaryListArgs, SummaryListItem } from '../../utils/summaryList'
 export default class SearchResultsView {
   constructor(private readonly presenter: SearchResultsPresenter) {}
 
+  private checkboxArgs({
+    name,
+    legend,
+    checkboxes,
+  }: {
+    name: string
```
Comment (Gweaton) [inline server/routes/findInterventions/searchResultsView.ts:13]: Probably one to add to the `govuk-frontend-types` that we introduced in a recent PR if we're keen to go down the route of strong types here, but we've got that covered in [this jira issue](https://dsdmoj.atlassian.net/browse/IC-1173?atlOrigin=eyJpIjoiODgwMDczYzBhZjQ1NGY5YWFhMmNjZjVmODdhYmM1MzUiLCJwIjoiaiJ9)

File: server/routes/findInterventions/searchSummaryPresenter.test.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+  describe('summary', () => {
+    describe('with an empty filter', () => {
+      it('returns no summary list items', () => {
+        const presenter = new SearchSummaryPresenter(new InterventionsFilter(), [])
+
+        expect(presenter.summary).toEqual([])
+      })
+    })
+
+    function linesForKey(key: string, args: [InterventionsFilter, PCCRegion[]]): string[] {
```
Comment (Gweaton) [inline server/routes/findInterventions/searchSummaryPresenter.test.ts:16]: I know the args types are slightly different here, but might be worth extracting the `linesForKey` function in `interventionDetailsPresenter` for this to some kind of "test helper" as this is very similar (on a quick glance, I could be wrong though...). Again, we can do this later, but if so, let's add a card for it.

**Review by Gweaton — APPROVED** [review]
Nice work 👏 

A few comments, but nothing serious, and probably stuff we can leave til after Show & Tell if you agree with any of it.

File: server/routes/findInterventions/interventionsFilter.ts (hunk 1 above)
Comment (lawrence-forooghian) [reply server/routes/findInterventions/interventionsFilter.ts:7]: It's not yet clear to me what the enum would represent. It's not "gender", it's just "the options that we display under ‘gender’ on the filter list". This type is quite specific to that particular set of filter options. If it becomes clear that those filter options need to be representable outside of one screen or something, then we can think about extracting them.

File: server/routes/findInterventions/searchResultsView.ts (hunk 2 above)
Comment (lawrence-forooghian) [reply server/routes/findInterventions/searchResultsView.ts:13]: Good point! I'll leave it until we work on that.

File: server/routes/findInterventions/interventionsFilter.ts (hunk 1 above)
Comment (Gweaton) [reply server/routes/findInterventions/interventionsFilter.ts:7]: Sorry, when I say "all over the codebase" I think I probably just meant this file (I saw a lot of 'male' and 'female' in the tests, so thought it'd make it more standardised, but it's probably not right to use this in the tests). Probably overkill, and you're right, it doesn't really represent "gender" itself.

File: server/routes/findInterventions/searchSummaryPresenter.test.ts (hunk 3 above)
Comment (lawrence-forooghian) [reply server/routes/findInterventions/searchSummaryPresenter.test.ts]: Good shout - [done](https://github.com/ministryofjustice/hmpps-interventions-ui/pull/128/commits/91b6079c96d90732621d3f52c3f6a075ab550798).

**Pull request merged**